src/machine/eval.rs

```diff
@@ -62,25 +60,35 @@ impl<W: Write, A: Arithmetic> Machine<W, A> {
             }
         } else {
             // If t1 is not a literal, evaluate it in place and return (if t1 then t2 else t3)
-            (self.step_in_place(t1.borrow_mut()), Term::Cond(t1, t2, t3))
+            let (changed, new_t1) = self.eval(*t1);
+            *t1 = new_t1;
+            (changed, Term::Cond(t1, t2, t3))
```

Comment:
This is a lot of duplicated code but we may be unable to put it in a function... so can we write a macro that does this evaluation? This may make the code more readable and shorter.

Reply:
I tried both to write a function and a macro and failed spectacularly.

Reply:
Ok that's sad I will try it locally.

Reply:
sure

Reply:
So I got something™️ that works.
```
macro_rules! eval_in_place {
    ($self:ident, $ty:ident, $val:ident ( $($arg:ident),* ) ) => {
        {
            let (changed, new_t) = $self.eval(*$ty);
            *$ty = new_t;
            (changed, $val( $( $arg, )* ))
        }
    };
}
```
It can be used like this
```
eval_in_place!(self, t1, App(t1, arg))
/*let (changed, new_t1) = self.eval(*t1);
t1 = new_t1;
changed, App(t1, arg))*/
```

Currently it requires the constructed type to be without a type path prefix so `App` and not `Term::App` but that should just need some more macro magic when we need it. Also self needs to be given explicitly.

Reply:
This is awesome. I had trouble making it work. Let me introduce it

Reply:
> This is awesome. I had trouble making it work. Let me introduce it

Wasn't easy to get the syntax right and easy to use 🤓.